libraries/lib-wave-track/WaveTrack.cpp

```diff
@@ -1652,6 +1464,54 @@ void WaveTrack::InsertClip(WaveClipHolder clip)
    mClips.push_back(std::move(clip));
 }
 
+void WaveTrack::ApplyStretchRatio(
+   std::optional<std::pair<double, double>> interval,
+   std::function<void(double)> reportProgress)
+{
+   assert(IsLeader());
+   if (GetNumClips() == 0)
+      return;
+   const auto startTime =
+      interval ? std::max(interval->first, GetStartTime()) : GetStartTime();
+   const auto endTime =
+      interval ? std::min(interval->second, GetEndTime()) : GetEndTime();
+   if (startTime == endTime)
+      return;
+   const auto numChannels = NChannels();
+   auto count = 0;
+   auto reportChannelProgress = [&](double progress) {
+      reportProgress((count + progress) / numChannels);
+   };
+   for (const auto pChannel : TrackList::Channels(this))
+   {
+      pChannel->ApplyStretchRatioOne(startTime, endTime, reportChannelProgress);
+      ++count;
+   }
+}
+
+void WaveTrack::ApplyStretchRatioOne(
+   double t0, double t1, const std::function<void(double)>& reportProgress)
+{
+   if (auto& clipAtT0 = *GetClipAtTime(t0);
+       clipAtT0.SplitsPlayRegion(t0) && clipAtT0.GetStretchRatio() != 1)
+      SplitAt(t0);
+   if (auto& clipAtT1 = *GetClipAtTime(t1);
+       clipAtT1.SplitsPlayRegion(t1) && clipAtT1.GetStretchRatio() != 1)
+      SplitAt(t1);
+   auto clip = GetClipAtTime(t0);
+   const auto numClips = GetNumClips(t0, t1);
+   auto count = 0;
+   auto reportClipProgress = [&](double progress) {
+      reportProgress((count + progress) / numClips);
+   };
+   while (clip && clip->GetPlayStartTime() < t1)
+   {
+      clip->ApplyStretchRatio(reportClipProgress);
+      clip = GetNeighbourClip(*clip, PlaybackDirection::forward);
```

Comment:
This loop repeatedly sorts the clips by time in `GetNeighbourClip`.

Avoid that by using `GetSortedClipArray` once at this level.